Initial draft of consensus messages.

matan-starkware opened the pull request.

matan-starkware committed:
Initial draft of consensus messages.

This is set up for Proposal streaming and voting. Certificates for a quorum are part of the sync protocol, not consensus. Peer state sharing is considered a future optimization.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,94 @@
+// Any content in the block itself, which is opaque to the consensus module (e.g. transactions,
+// proofs, gas price, etc.)
+message ProposalContent {
+    bytes              chunk     = 1;
```
We know what we're proposing, no need to make it opaque.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,94 @@
-// A block proposal is a series of (Transactions+, StateDiff)* BlockHeader
+// Digest of the block, which represents the streamed content (e.g. StateDiffRoot, ProofRoot).
+// Opaque to the consensus module.
+message Commitment {
```
There are already types for Hash, Merkle and Patricia. No need to introduce a new one.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,94 @@
+    bytes data = 1;
+}
+
+// Data which is opaque to the consensus module, but must come with the ProposalInit to allow the
```
No need to make it opaque.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,94 @@
-        Prevote   = 2;
-        Precommit = 3;
+        Prevote   = 0;
+        Precommit = 1;
```
The unknown and proposal votes were taken from cometbft, why can we do without them?

matan-starkware replied:
I guess we should ask them, but my thinking was:

A proposal is its own message. Calling a Vote "Proposal" is meaningless to me. And Unknown also is a meaningless type, it just isn't a valid vote, so why give the option to send an invalid vote?

The way I though about this field was just as a way to make sure the Prevote/Precommit end up with different signatures.

ittaysw replied:
I guess that when a proposer creates a proposal they also create a vote for it.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,94 @@
+    // votes have the same fields.
+    Type               type         = 1;
+    uint64             fork_id      = 2;
+    uint64             block_number = 3;
```
If we have BlockNumber, we should use it here.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,94 @@
+message ProposalInit {
+    BlockNumber               block_number = 1;
+    uint32                    round        = 2;
+    PublicKey                 proposer     = 3;
```
Incentives will go to an address, this is why I used it and not a public key. That is the ID. The other validators can associate the address with its public key (e.g. , read from the blockchain)

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,94 @@
+    // Hash of all fields that make up the block.
+    Hash               block_hash       = 2;
+    // Proposer's signature of `block_hash`.
+    ConsensusSignature block_signature  = 3;
```
This is essentially a proposal vote (which is treated like a prevote)

matan-starkware replied:
I think this adds complexity and I don't see why. I'd prefer for all nodes to just send prevotes in the normal manner.

ittaysw replied:
The comment was about the different treatment of the proposer vote vs all other votes. It can be the same.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,94 @@
+    uint64             fork_id      = 2;
+    uint64             block_number = 3;
+    uint32             round        = 4;
+    Hash               block_hash   = 5;
```
Should be optional to allow a 'nil' vote

matan-starkware replied:
Happy to make optional explicit, but since `Hash` is a user defined type, in proto3 it is automatically optional.

ittaysw replied:
They have revised this decision in proto3 3.15. So you need to mark as optional

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,94 @@
+// Commitment or just hash?
+message ProposalFin {
+    // Identifies the content of the block that was streamed.
+    Commitment         commitment       = 1;
```
how is 'commitment' different than 'block_hash'?

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,94 @@
+    PublicKey                 proposer     = 3;
+    ProposalMetadata          metadata     = 4;  // Needed???
+    // The hash of all the above fields.
+    Hash                      hash         = 5;
```
Not required. The validators will hash.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,94 @@
+    // Identifies the content of the block that was streamed.
+    Commitment         commitment       = 1;
+    // Hash of all fields that make up the block.
+    Hash               block_hash       = 2;
```
Not required, the validators will hash

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
-    google.protobuf.Timestamp timestamp         = 6;
-    ConsensusSignature        signature         = 7;
+message Transaction {
+    bytes elements = 1;
```
There's already a transaction type in common

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
+}
+
+message TransactionBatch {
+    repeated Transaction transactions = 1;
```
There's already a Transactions object.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
+}
+
+message Proof {
+    bytes elements = 1;
```
There's already a Proof type

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
+message BlockNumber {
+    Hash   parent_hash = 1;
+    uint64 height      = 2;
+    uint64 fork_id     = 3;
```
What does fork_id mean? What is the value?

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
+}
+
+message BlockNumber {
+    Hash   parent_hash = 1;
```
Why do we need this in the block number? Why isn't block_number === height?

matan-starkware replied:
I'll include the fork_id question here. BlockNumber is about identifying where in the blockchain a block lies.

I agree that `height` and `parent_hash` are basically equivalent. I just thought it was nice to have them both explicitly.

`fork_id` though is used in the case that there is a fork, when the L1 contract resets the validator set. If the L1 contract has finalized Starknet block `H` when it forks, it is possible that Starknet already decided within L2 block `H+1`. Starknet now needs to return to height `H+1` because of the L1 fork. We need a way to disambiguate the `H+1` from before the fork and the `H+1` from after the fork.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
-// A block proposal is a series of (Transactions+, StateDiff)* BlockHeader
+message Voter {
+    Address address = 1;
+    uint64  weight  = 2;
```
Why do we need to pass weight in p2p?

matan-starkware replied:
Well since we don't pass the validator set in the proposal anymore we don't need this at all.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
+    // messages, to make sure the data, and therefore the signature, are unambiguous, since the
+    // votes have the same fields.
+    Type               type         = 1;
+    uint64             fork_id      = 2;
```
so is fork_id here or in BlockNumber?

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
+    BlockNumber               block_number   = 1;
+    uint32                    proposal_round = 2;
+    uint32                    valid_round    = 3;
+    Address                   proposer       = 4;
```
We need a signature as well. Otherwise malicious actors may send conflicting messages.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
+// 1. The first message is `init`.
+// 2. The next N messages are `content`.
+// 3. The last message is `fin`.
+message Proposal {
```
Better make this a ProposalMessage and then a ProposalPart has repeated ProposalMessage to fit several in one protobuf

matan-starkware replied:
Done, but I'm not 100% confident I understand why. Is this a small proposal optimization? Meaning if the proposal is small enough then we can fit all of the parts into a single protobuf message?

If so I feel like the old structure is simpler, you get a single "part" per message", and it's optimized for the case we care about, large proposals where a TxBatch fills up the proto.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
+// A proposal is a stream:
+// 1. The first message is `init`.
+// 2. The next N messages are `content`.
+message Certificate {
```
why does a certificate contain the entire proposal? It should just be the block hash

matan-starkware replied:
Yeah this is a question of both what we want and naming. In some cases a node may already have a proposal but doesn't have all the votes to recognize a quorum. In other cases a node may also not even have received the proposal. Should I create 2 types of messages? Certificates and ???

ittaysw replied:
A certificate is after everyone signed and we have a bonafide block. A node can just ask it of its peers.

matan-starkware replied:
Yes exactly. The certificate effectively sends:
1. ProposalInit + ProposalFin + Votes
2. Stream of content
If a receiver has already received this proposal content it will just drop the stream. If it has yet to receive the proposal it needs to before it will accept this decision.

ittaysw replied:
I'm saying it's the other way round. You have a proposal, then everyone signs on it. When there are enough signatures, that is a certificate and it is part of the block header. It is just a bunch of signatures.

matan-starkware replied:
Removed certificates from here:
- Prevote - premature optimization
- Precommit - this is in the sync protocol, not consensus

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
+    // Votes of 1 type which form a quorum in favor of the proposal.
+    repeated Vote      votes            = 2;
+    // Proposer's signature of `block_hash`.
+    ConsensusSignature block_signature  = 3;
```
This can be just one of the votes

matan-starkware replied:
I'm not sure I understand. The `block_signature` is part of the proposal. The `ProposalFin` is sent at the end since the proposal is built in parallel with the streaming. For a certificate though we already know the proposal and can just send the `block_signature` at the beginning.

I think you are saying that we can consider this equivalent to the Proposer's vote in favor of the proposal? But I think that is being too clever and only saves us a few bytes. I like the simpler layout (at least for small non-repeated fields).

ittaysw replied:
The signature of the proposer on the block can be thought of as being just like any other vote on the block.

matan-starkware replied:
I think I understand the ideal here, but I believe it is simpler to just send proposals and votes separately at stick to messages looking like the description in the paper. I feel like this is an optimization for something that is not expensive, meaning a single vote.

ittaysw replied:
A proposer sends a proposal with their signature. When a certificate is produced, that signature can just be another one

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
+    }
+}
+
+message PeerState {
```
when do peers exchange this?

matan-starkware replied:
I was basically just planning to take from comet where peers share their state in a push manner whenever any of these fields is updated. (That's how peers know if your lagging for instance)

ittaysw replied:
So this is not part of the consensus. Please take it out.

matan-starkware replied:
It sort of is part of the consensus. Meaning, consensus assumes that after GST all messages, including from pre-GST arrive. This is not realistic. PeerState lets nodes know when a neighbor is lagging and to send it the votes it was missing. It's a part of converting consensus to a realistic protocol.

ittaysw replied:
What message is it that sends missing votes? 
Will peers constantly send PeerState to other peers? This needs to be at a high rate since the entire round takes 1-2 seconds to vote. If a peer is lagging, will its state even arrive to other nodes (there's a reason it is lagging)? 

In my mind, a lagging peer will simply not vote. 

I don't think that consensus assumes that pre-GST messages arrive after GST

matan-starkware replied:
Tendermint is built on the partially synchronous model which does assume that pre-GST messages arrive after GST.

I think we can remove PeerState sharing now and come back to it in the future.

matan-starkware replied:
Removed

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
-        StateDiff    state_diff   = 2;
-        Proposal     proposal     = 3;
+        ProposalInit    init    = 1;
+        ProposalContent content = 2;
```
Since ProposalContent is just a oneof message, maybe just put TransactionBatch and Proof here.

matan-starkware replied:
I added content since you pointed out it should be signed.

ittaysw replied:
And ProposalInit is also signed. So everything is signed and can be treated the same.

matan-starkware replied:
I thought we had discussed and that only ProposalFin needed a signature? From the idealized perspective of consensus that is certainly true, as the signed fin `(height, round, block_hash)` is enough to identify the proposal.

Are you worried that a malicious node will attempt to grief us by sending proposal's when it is not its turn and so we want a signature on the init to make sure we only accept proposals for a node from the correct round?

ittaysw replied:
If an attacker sends their own ProposalContent, how can peers know it is not from the proposer? If peers use it then they end up with a block hash that is different than the proposer's. That's a DoS attack.

ittaysw replied:
Note that the signature at fin is about the entire block (hash of the header), and plays the role of a vote while the previous ones are just for the messages authenticity.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,121 @@
+    bytes elements = 1;
+}
+
+message TransactionBatch {
```
Shouldn't this and Proof contain signatures? Otherwise, an attacker can send these and then validators will end up with a differnet hash than the proposer.

matan-starkware replied:
Oh as a protection against a malicious validator "replaying" the proposal but lying about the content. Interesting.

matan-starkware replied:
On second thought I am not certain we need this. If a malicious validator is meant to relay a proposal we are worried about:
1. Sending invalid content - the signature on `ProposalContent` protects against this.
2. Censoring content - the signature on `ProposalContent` does not protect against this.

The `block_signature` in the `ProposalFin` message should protect us against both of these scenarios. In which case the signature on `ProposalContent` is unneeded for the proposal as a whole. 

Maybe it is useful since a malicious validator cannot waste our resources with invalid `ProposalContent` (DoS attack), but they could still do other things to achieve a similar end (many proposals in parallel). I'm leaning back towards removing this and just keeping the signature on the proposal as a whole.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,115 @@
-    google.protobuf.Timestamp timestamp         = 6;
-    ConsensusSignature        signature         = 7;
+
+message BlockNumber {
```
I think calling it block *number* when it contains also a hash and fork is confusing.

matan-starkware replied:
What do you think of BlockLocation?

The point of this field is to tell you where this block belongs in the chain. I also don't want anything like BlockId since these fields tell you nothing about the content of the block.

matan-starkware replied:
The other option is just to flatten this struct and put the fields directly into the messages without this encapsulation.

BTW we will need to update BlockHeader to include `fork_id` .

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -9,107 +9,146 @@ import "google/protobuf/timestamp.proto";
     uint32                    proposal_round = 2;
     optional uint32           valid_round    = 3;
     Address                   proposer       = 4;
-    google.protobuf.Timestamp timestamp      = 5;
```
removed the timestamp?

matan-starkware replied:
Yes as discussed this is content, not part of consensus

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -53,27 +53,29 @@ message ProposalInit {
+        // risk a scenario where in round `R` proposal P is locked. Then in a later round the
+        // timestamp in P has gone stale. Therefore the lower bound should be "greater than the
+        // previous timestamp". Upper bounds don't suffer from this problem.
+        BlockMetadata metadata            = 1;
```
Where is BlockMetadata defined? 

Also, it may be confusing if header.proto has messages that are different than here (e.g. BlockLocation is only here)

matan-starkware replied:
It seems I hallucinated this, but let me describe my thoughts. A block is made of: state diff (built from TX list), proof, metadata (max cairo steps, gas price, height, parent hash, etc.). Basically we already have dedicated fields for TX batches and the proof, so I thought we needed another field for all of the other info that is sent.

Where should we define this? Do you have a different name you prefer?

matan-starkware replied:
Discussed, no current usage so removed.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -53,27 +53,29 @@ message ProposalInit {
     ConsensusSignature signature          = 5;
 }
 
 message ProposalFin {
```
It may be confusing to others that Proposal is init, content, fin where content has more "stuff". maybe flatten?

matan-starkware replied:
I made the change but I want to point some things out which give me pause:
- I'm not certain it's a problem for there to be more stuff within. Have we limited other messages due to it holding yet further messages? (I agree this is an enum to an enum so a bit different)
- It ties the message's structures. The thing I really notice is that for `ProposalFin` we want there to be the a proper Tendermint signature on the Proposal. Given the flattened structure though I had to move `sequence_number` and `signature` into the `Proposal` messaage. So I end up with another `signature` field inside of `ProposalFin`.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,80 @@
-    google.protobuf.Timestamp timestamp         = 6;
-    ConsensusSignature        signature         = 7;
+
+message BlockLocation {
```
I think these messages should be consistent with the state sync.

matan-starkware replied:
I flattened it like we discussed in person. I know `SignedBlockHeader` hash `parent_hash` and `number` (I used `block_number`). No one uses fork_id yet, as that is a concept we added recently.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,80 @@
+// `R` proposal P is locked. Then in a later round the timestamp in P has gone stale. Therefore the
+// lower bound should be "greater than the previous timestamp". Upper bounds don't suffer from this
+// problem.
+message Proposal {
```
We also need to pass Vote messages on the same topic.

matan-starkware replied:
I created the ConsensusMessage proto below.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,80 @@
+    ConsensusSignature signature      = 8;
+}
+
+message ProposalInit {
```
Note that after unifying proposal and vote messages, proposal init looks a lot like a vote.

matan-starkware replied:
They certainly have some similar fields, but I fear that it would just make the protobuf source code harder to understand if we used the same struct. As in ProposalInit becomes Vote but with `step=Propose` (instead of VoteType) and `block_hash=None`, it seems weird to call the struct `Vote`. But yes we could do this if you feel strongly.

We'd also have some Vote wrapper to move the signature there since ProposalInit doesn't itself have signature.

ittaysw replied:
But the Proposal message does. Anyway, I don't feel strongly about it.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -2,41 +2,84 @@ syntax = "proto3";
-    int32              validator_index   = 3;  // ???
-    ConsensusSignature signature         = 4;
+// Create a wrapper for a list of transactions, as we cannot use `repeated` in a oneof.
+message Transactions {
```
We have this elsewhere.

matan-starkware replied:
Re-added to transaction.proto

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -2,41 +2,84 @@ syntax = "proto3";
+}
+
+// TODO: Define the actual block proof message in another file.
+message BlockProof {}
```
Isn't this in header.proto?

matan-starkware replied:
It was deleted. I readded it there now.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -2,41 +2,84 @@ syntax = "proto3";
+    ConsensusSignature signature = 5;
 }
+
+message ConsensusMessage {
```
Don't we want this to contain the signature (so all consensus messages are always signed)?

matan-starkware replied:
Moved the signature fields to here.

ittaysw replied:
But This signature can be x86 friendly. Just the precommit signature needs to be stark friendly, so they can be checked by the OS.

matan-starkware replied:
I'm not sure what you are asking for.
1. I would have thought that we wanted all consensus signatures to be of the same format (ie in the precommit, prevote, and ProposalFin). Only the parts of the proposal (init, content) would use a different signature scheme.
2. If we have multiple signature schemes how are we planning to get the public keys to the consensus component in each node?
3. What would you want changed in the proto based on the different schemes? I presumed this is something that the receiving node handles based on the message it received (see point 1).

ittaysw replied:
1. but now you are using the consensus signature on the proposal as well (field 3 below)
2. you get multiple public keys in the same way you get a single public key.
3. I think that Precommit votes have stark signatures, all the other signatures can be x86. 

Note for context: singing with a stark curves takes (IIRC) 100ms in a cloud instance. That is a lot of time. I think verification may even be more costly and in the prevote will be done on at least 67 votes.

matan-starkware replied:
Your initial request was for "this to contain the signature". I have moved the signature here. What change you are asking for?

matan-starkware replied:
Spoke f2f:
- Precommit - Stark friendly signature, as this must be validated in StarknetOS
- Other - x86 friendly signature for performance
The structure of ConsensusMessage does not change though.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,75 @@
+// Produce the full Tendermint Proposal message once the proposer node has finished building it.
+// Tendermint just ID's the proposal's content, block_hash.
+message ProposalFin {
+    Hash            parent_hash    = 1;
```
If ProposalFin needs to repeat fields form ProposalInit, why don't Transactions and BlockProof need to have them?

ittaysw replied:
Either this is taken care at the protocol level, or at the message wrapper. Have a 'proposal id' (that may be a hash of the proposal init) which is added to every message and is signed on.

matan-starkware replied:
The reason ProposalFin repeats the field is because we need the proposer to sign on all of these fields. I always want the signature on a message to be only on the content in the message. If I don't repeat fields then the consensus signature is on some fields from ProposalFin and some from ProposalInit. Updated the comment.

I think the structure here is more straightforward than adding a new "proposal id". The Proposal struct from Tendermint is well defined and we aren't expecting it to change.

ittaysw replied:
So, you can leave fields out of ProposalFin and have the signature includes those from the proposalinit, OR, you need to repeat fields in each message including transactions and proof. I don't think the latter is required.
Clients would still need to make sure the fields in ProposalFin match those in ProposalInit, so what is the added value in repeating them? 
Note you can hash those and then following messages include the hash and sign on it.

matan-starkware replied:
For what I want to achieve, no I do not need to repeat the fields in the context (transactions & proof). I want to have the ProposalFin include the full consensus signature on the proposal. I want the signature to match the content of the message. Therefore, only in ProposalFin do I need to repeat.

ittaysw replied:
Don't repeat. The proposer needs to sign on these, but no need to repeat. Nodes will need to keep the init message and confirm the fields are the same, so the proposer is not cheating (telling them to validate against one state and then finalizing as if it's another). 

And, the signature will need to include this information for transactions and proof as well to support message interleaving (in the rare case that a proposer proposes on several rounds and there are network issues so messages come out of band, especially since we user relays.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,73 @@
+    uint64  block_number   = 2;
+    uint64  fork_id        = 3;
+    uint32  proposal_round = 4;
+    Address proposer       = 5;
```
we know who the proposer is.

ittaysw replied:
also parent_hash.

ittaysw commented on p2p/proto/consensus.proto:
```diff
@@ -1,42 +1,73 @@
+message ProposalFin {
+    optional uint32 valid_round    = 1;
+    // Validators must verify this matches the block_hash they calculated from the content.
+    Hash            block_hash     = 2;
```
we don't need this, it's derived by each client from the information they received.

matan-starkware committed:
Remove parent_hash as we already have (block_number, fork_id).

Also remove block_hash from ProposalFin. The fear is that some validators will mistakenly accept this instead of calculating and confirming it on their own.
